### PR TITLE
Default branch option (nothing set) checkout to default branch on remote

### LINKED ISSUE #92: Default remote branch targeting master
## Problem

When cloning with non defined branch the operation points to the `master` branch, instead of the default one on git. This was not a problem before as most of the packages had the `master` as default, however today, git is changing to `main` as discussed [here](https://github.com/github/renaming). The command issued on log with clonning is the following:

```bash
git --git-dir ${PATH}.git fetch git@github.com:${GIT_REPO}.git refs/heads/master
```

## Analysis 
As the sources are loaded, the `remote_branch` member is loaded with the following values, as can be seen [here](https://github.com/rock-core/autobuild/blob/1f55666d28944603b257c6969c55c814aae8a686/lib/autobuild/import/git.rb#L1214)
```ruby
remote_branch =
                options[:remote_branch] || options[:branch] ||
                self.remote_branch || 'master'
```
And the clone follows this ref.

## Proposition
When there is no branch specified, do not specify remote_branch as well (maybe using an empty string?)

What you guys think, am I missing some potential problem?

### PR DESCRIPTION
**Pull request by ivan1993br**
Sorry, for the delay guys,

# Description

This PR set the default branch option for git remote to the current default branch on git repo

The logic is the following:

- If no branch is given on source definition for git packages:
   - Check local git for the remote HEAD definition
   - If it does not exists, check remote git

An obs: the "plain" clone happens before the "default" branch identification, so the git remote repo check will happens mostly on existent branch with no remote HEAD

# Issue
closes #92

### CODE REVIEW

File: lib/autobuild/import/git.rb [hunk 1]
```diff
@@ -293,6 +293,52 @@ def validate_importdir(package)
             git_dir(package, true)
         end
 
+        
+        # Return default local branch if exists, if not return nil
+        #
+        # @param [Package] package
+        def default_local_branch(package)
```
Comment (doudou) [inline lib/autobuild/import/git.rb:300]: This name is misleading. It does not returnthe default local branch, it returns the default remote branch as stored locally.

Maybe `try_resolve_remote_head_from_local_remotes`. `default_remote_branch` would become `try_resolve_remote_head_from_ls_remote`. Finally `default_branch` would become `resolve_remote_head`

File: lib/autobuild/import/git.rb [hunk 2]
```diff
@@ -293,6 +293,52 @@ def validate_importdir(package)
...
+        end
+
+        # Return default remote branch if exists, if not return 'master'
+        #
+        # @param [Package] package
+        def default_remote_branch(package)
+            ls_remote_string = package.run(:import,
+                Autobuild.tool('git'), 'ls-remote', '--symref', repository).first.strip
+            default_branch_match = ls_remote_string.match("ref:[^A-z]refs/heads/(.*)[^A-z]HEAD")
+            if !default_branch_match.nil?
```
Comment (doudou) [inline lib/autobuild/import/git.rb:315]: No need to explicitely test against nil ... just do

~~~
if default_branch_match
    default_branch_match[1]
else
    "master"
end
~~~

File: lib/autobuild/import/git.rb [hunk 3]
```diff
@@ -293,6 +293,52 @@ def validate_importdir(package)
...
+                return 'master'
+            end
+        end
+
+        # Return default local branch if exists, if not return the default remote branch
+        #
+        # @param [Package] package
+        def default_branch(package)
+            local_branch = default_local_branch(package)
+            if local_branch.nil?
```
Comment (doudou) [inline lib/autobuild/import/git.rb:327]: Same remark about the explicit `nil` test. You don't do that in Ruby.

File: lib/autobuild/import/git.rb [hunk 4]
```diff
@@ -706,18 +755,24 @@ def has_branch?(package, branch_name)
         end
 
         def has_local_branch?(package)
-            has_branch?(package, local_branch)
+            if !local_branch.nil?
+                has_branch?(package, local_branch)
+            else
+                false
+            end
```
Comment (doudou) [inline lib/autobuild/import/git.rb:762]: This is simply

~~~
has_branch?(...) if local_branch
~~~

File: lib/autobuild/import/git.rb [hunk 5]
```diff
@@ -706,18 +755,24 @@ def has_branch?(package, branch_name)
...
         def detached_head?(package)
             current_branch(package).nil?
         end
 
         private def remote_branch_to_ref(branch)
-            if branch.start_with?("refs/")
-                branch
-            else
-                "refs/heads/#{branch}"
+            if !branch.nil?
```
Comment (doudou) [inline lib/autobuild/import/git.rb:770]: Do a guard and then the rest

~~~
return unless branch

if branch.start_with?(...)
...
end
~~~

File: test/import/test_git.rb [hunk 6]
```diff
@@ -741,3 +748,114 @@ def import(**options)
         importer.import(pkg)
     end
 end
+
+describe Autobuild::Git do
```
Comment (doudou) [inline test/import/test_git.rb:752]: This block should not be repeated. Just add the new describe blocks to the existing toplevel `describe` block.

File: lib/autobuild/import/git.rb [hunk 7]
```diff
@@ -293,6 +293,52 @@ def validate_importdir(package)
...
+        def default_branch(package)
+            local_branch = default_local_branch(package)
+            if local_branch.nil?
+                return default_remote_branch(package)
+            else
+                return local_branch
+            end
+        end
+
+        def has_all_branches?
```
Comment (g-arjones) [inline lib/autobuild/import/git.rb:334]: This is just

```ruby
def has_all_branches?
    remote_branch && local_branch
end
```

File: lib/autobuild/import/git.rb [hunk 8]
```diff
@@ -477,6 +523,9 @@ def self.run_git_bare(package, *args)
         #
         # Set a remote up in the repositorie's configuration
         def setup_remote(package, remote_name, repository, push_to = repository)
+            if !has_all_branches?
```
Comment (g-arjones) [inline lib/autobuild/import/git.rb:526]: Use `unless` rather than negative condition.

```ruby
unless has_all_branches
    relocate(repository, default_branch: default_branch(package))
end
```

File: lib/autobuild/import/git.rb [hunk 9]
```diff
@@ -1065,6 +1120,11 @@ def determine_target_state(package, only_local: false)
         # @option (see Package#update)
         def update(package, options = Hash.new)
             validate_importdir(package)
+
+            if !has_all_branches?
```
Comment (g-arjones) [inline lib/autobuild/import/git.rb:1124]: Avoid negative conditions...

File: lib/autobuild/import/git.rb [hunk 10]
```diff
@@ -1210,13 +1270,15 @@ def relocate(repository, options = Hash.new)
...
                 options[:remote_branch] || options[:branch] ||
-                self.remote_branch || 'master'
-            if local_branch.start_with?("refs/")
-                raise ArgumentError, "you cannot provide a full ref for"\
-                  " the local branch, only for the remote branch"
+                self.remote_branch || options[:default_branch] || nil
+            if local_branch
+                if local_branch.start_with?("refs/")
+                    raise ArgumentError, "you cannot provide a full ref for"\
+                    " the local branch, only for the remote branch"
```
Comment (g-arjones) [inline lib/autobuild/import/git.rb:1280]: Indentation

File: test/import/test_git.rb [hunk 11]
```diff
@@ -26,11 +26,11 @@
         it "takes a local_branch argument" do
             git = Autobuild::Git.new('repo', local_branch: 'test')
             assert_equal "test", git.local_branch
-            assert_equal "master", git.remote_branch
+            assert_equal nil, git.remote_branch
```
Comment (g-arjones) [inline test/import/test_git.rb:29]: `assert_nil`

File: lib/autobuild/import/git.rb [hunk 12]
```diff
@@ -293,6 +293,52 @@ def validate_importdir(package)
...
 
+        
+        # Return default local branch if exists, if not return nil
+        #
+        # @param [Package] package
+        def default_local_branch(package)
+            ls_local_string = run_git(package, 'symbolic-ref', "refs/remotes/#{@remote_name}/HEAD").first.strip
+            default_local_branch = ls_local_string.match("refs/remotes/#{@remote_name}/(.*)")
+            default_local_branch.nil? ? nil : default_local_branch[1]
+            rescue Autobuild::SubcommandFailed
```
Comment (g-arjones) [inline lib/autobuild/import/git.rb:304]: `rescue` should have the same indentation level as `def` 

(also the explicit `return` is not necessary)

**Review by g-arjones — CHANGES_REQUESTED** [review]
The implementation looks good... Just some style issues (most of them would have been spotted by rubocop)

File: lib/autobuild/import/git.rb (hunk 7 above)
Comment (ivan1993br) [reply lib/autobuild/import/git.rb:334]: A curiosity, should a boolean method return `nil` instead of false in ruby? I am asking this because `nil && nil => nil`.

Comment (g-arjones) [reply lib/autobuild/import/git.rb:334]: In Ruby `nil` and `false` are the only things that evaluates to falsy (even `0` is truthy). So, in boolean contexts they mean the same.

**Comment by ivan1993br** [comment]
Thanks for the feedback, all requested modification were implemented

**Review by doudou — APPROVED** [review]

**Review by g-arjones — APPROVED** [review]
